File: src/common/vectordb.py
"""Vector store client with two backends behind one interface.

Production Aurora is reached via the RDS Data API (no VPC needed for any
Lambda in this pipeline -- see infra/cloud/modules/aurora's comments). The
Data API has no local emulator, so local docker-compose talks to the plain
`pgvector/pgvector` container over the normal Postgres wire protocol via
psycopg instead. Toggle with DB_MODE=data_api|psycopg (default data_api,
matching the EMBEDDINGS_PROVIDER=fake toggle already used for Bedrock).

Both backends run equivalent SQL against the same `chunks` table (see
infra/local/postgres-init/002-schema.sql / infra/cloud/modules/aurora's
schema null_resource, which must stay identical).
"""

from __future__ import annotations

import json
import os

import boto3

DB_MODE = os.environ.get("DB_MODE", "data_api")

_UPSERT_COLUMNS = """
    INSERT INTO chunks (chunk_id, file_id, text, embedding, metadata)
    VALUES ({placeholders})
    ON CONFLICT (chunk_id) DO UPDATE SET
      file_id = EXCLUDED.file_id,
      text = EXCLUDED.text,
      embedding = EXCLUDED.embedding,
      metadata = EXCLUDED.metadata
"""


def _vector_literal(vector: list[float]) -> str:
    """pgvector accepts a JSON-array-shaped text literal, e.g. '[0.1,0.2]'."""
    return json.dumps(vector)


class VectorDBClient:
    """One instance per Lambda invocation -- open at the top of the handler,
    `close()` in `finally`, mirroring "one Aurora connection per invocation"
    (meaningful for the psycopg backend; the Data API backend is a stateless
    HTTPS client, so close() is a no-op there)."""

    def __init__(self) -> None:
        self._mode = DB_MODE
        if self._mode == "psycopg":
            self._conn = _connect_psycopg()
        else:
            self._client = boto3.client("rds-data")
            self._cluster_arn = os.environ["AURORA_CLUSTER_ARN"]
            self._secret_arn = os.environ["AURORA_SECRET_ARN"]
            self._database = os.environ["AURORA_DATABASE_NAME"]

    def upsert_chunk(
        self,
        chunk_id: str,
        file_id: str,
        text: str,
        vector: list[float],
        metadata: dict,
    ) -> None:
        if self._mode == "psycopg":
            self._upsert_psycopg(chunk_id, file_id, text, vector, metadata)
        else:
            self._upsert_data_api(chunk_id, file_id, text, vector, metadata)

    def _upsert_psycopg(
        self, chunk_id: str, file_id: str, text: str, vector: list[float], metadata: dict
    ) -> None:
        sql = _UPSERT_COLUMNS.format(placeholders="%s, %s, %s, %s::vector, %s::jsonb")
        with self._conn.cursor() as cur:
            cur.execute(
                sql,
                (chunk_id, file_id, text, _vector_literal(vector), json.dumps(metadata)),
            )
        self._conn.commit()

    def _upsert_data_api(
        self, chunk_id: str, file_id: str, text: str, vector: list[float], metadata: dict
    ) -> None:
        sql = _UPSERT_COLUMNS.format(
            placeholders=(
                ":chunk_id, :file_id, :text, "
                "CAST(:embedding AS vector), CAST(:metadata AS jsonb)"
            )
        )
        self._client.execute_statement(
            resourceArn=self._cluster_arn,
            secretArn=self._secret_arn,
            database=self._database,
            sql=sql,
            parameters=[
                {"name": "chunk_id", "value": {"stringValue": chunk_id}},
                {"name": "file_id", "value": {"stringValue": file_id}},
                {"name": "text", "value": {"stringValue": text}},
                {"name": "embedding", "value": {"stringValue": _vector_literal(vector)}},
                {"name": "metadata", "value": {"stringValue": json.dumps(metadata)}},
            ],
        )

    def close(self) -> None:
        if self._mode == "psycopg":
            self._conn.close()


def _connect_psycopg():
    import psycopg

    return psycopg.connect(
        host=os.environ.get("POSTGRES_HOST", "localhost"),
        port=int(os.environ.get("POSTGRES_PORT", "5432")),
        dbname=os.environ.get("POSTGRES_DB", "conduit_rag"),
        user=os.environ.get("POSTGRES_USER", "conduit_rag"),
        password=os.environ.get("POSTGRES_PASSWORD", "conduit_rag"),
    )

File: frontend/lib/vectordb.ts
import { RDSDataClient, ExecuteStatementCommand } from "@aws-sdk/client-rds-data";
import { Client as PgClient } from "pg";

// Aliased: `pg`'s own query() return value is also structurally a
// "QueryResult" -- this import is our app-level chunk-result shape instead.
import type { QueryResult as ChunkResult } from "./types";

/**
 * Mirrors src/common/vectordb.py's DB_MODE toggle: production Aurora is
 * reached via the RDS Data API (no VPC needed), which has no local emulator,
 * so local docker-compose talks to the plain `pgvector/pgvector` container
 * over the wire protocol via `pg` instead. Both run against the same
 * `chunks` table (infra/local/postgres-init/002-schema.sql /
 * infra/cloud/modules/aurora).
 */
const DB_MODE = process.env.DB_MODE ?? "data_api";

export async function deleteChunksByFileId(fileId: string): Promise<void> {
  if (DB_MODE === "psycopg") {
    await deletePsycopg(fileId);
  } else {
    await deleteDataApi(fileId);
  }
}

async function deletePsycopg(fileId: string): Promise<void> {
  const client = new PgClient({
    host: process.env.POSTGRES_HOST ?? "localhost",
    port: Number(process.env.POSTGRES_PORT ?? "5432"),
    database: process.env.POSTGRES_DB ?? "conduit_rag",
    user: process.env.POSTGRES_USER ?? "conduit_rag",
    password: process.env.POSTGRES_PASSWORD ?? "conduit_rag",
  });
  await client.connect();
  try {
    await client.query("DELETE FROM chunks WHERE file_id = $1", [fileId]);
  } finally {
    await client.end();
  }
}

async function deleteDataApi(fileId: string): Promise<void> {
  const client = new RDSDataClient({});
  await client.send(
    new ExecuteStatementCommand({
      resourceArn: process.env.AURORA_CLUSTER_ARN,
      secretArn: process.env.AURORA_SECRET_ARN,
      database: process.env.AURORA_DATABASE_NAME,
      sql: "DELETE FROM chunks WHERE file_id = :file_id",
      parameters: [{ name: "file_id", value: { stringValue: fileId } }],
    }),
  );
}

/** pgvector accepts a JSON-array-shaped text literal, e.g. '[0.1,0.2]' -- mirrors common/vectordb.py's _vector_literal. */
function vectorLiteral(vector: number[]): string {
  return JSON.stringify(vector);
}

// Cosine distance via the chunks_embedding_hnsw_idx (vector_cosine_ops)
// index created in 002-schema.sql -- `<=>` returns distance, so similarity
// is 1 minus that.
const SIMILARITY_SEARCH_SQL = `
  SELECT chunk_id, file_id, text, metadata, 1 - (embedding <=> {embedding}) AS similarity
  FROM chunks
  ORDER BY embedding <=> {embedding}
  LIMIT {top_k}
`;

export async function similaritySearch(vector: number[], topK: number): Promise<ChunkResult[]> {
  if (DB_MODE === "psycopg") {
    return searchPsycopg(vector, topK);
  }
  return searchDataApi(vector, topK);
}

async function searchPsycopg(vector: number[], topK: number): Promise<ChunkResult[]> {
  const client = new PgClient({
    host: process.env.POSTGRES_HOST ?? "localhost",
    port: Number(process.env.POSTGRES_PORT ?? "5432"),
    database: process.env.POSTGRES_DB ?? "conduit_rag",
    user: process.env.POSTGRES_USER ?? "conduit_rag",
    password: process.env.POSTGRES_PASSWORD ?? "conduit_rag",
  });
  await client.connect();
  try {
    const sql = SIMILARITY_SEARCH_SQL.replace(/{embedding}/g, "$1::vector").replace(
      "{top_k}",
      "$2",
    );
    const result = await client.query(sql, [vectorLiteral(vector), topK]);
    return result.rows.map((row) => ({
      chunk_id: row.chunk_id,
      file_id: row.file_id,
      text: row.text,
      metadata: row.metadata,
      similarity: Number(row.similarity),
    }));
  } finally {
    await client.end();
  }
}

async function searchDataApi(vector: number[], topK: number): Promise<ChunkResult[]> {
  const client = new RDSDataClient({});
  const sql = SIMILARITY_SEARCH_SQL.replace(
    /{embedding}/g,
    "CAST(:embedding AS vector)",
  ).replace("{top_k}", ":top_k");
  const response = await client.send(
    new ExecuteStatementCommand({
      resourceArn: process.env.AURORA_CLUSTER_ARN,
      secretArn: process.env.AURORA_SECRET_ARN,
      database: process.env.AURORA_DATABASE_NAME,
      sql,
      parameters: [
        { name: "embedding", value: { stringValue: vectorLiteral(vector) } },
        { name: "top_k", value: { longValue: topK } },
      ],
      formatRecordsAs: "JSON",
    }),
  );
  const rows: Array<{
    chunk_id: string;
    file_id: string;
    text: string;
    metadata: Record<string, unknown>;
    similarity: number;
  }> = JSON.parse(response.formattedRecords ?? "[]");
  return rows.map((row) => ({
    chunk_id: row.chunk_id,
    file_id: row.file_id,
    text: row.text,
    metadata: row.metadata,
    similarity: row.similarity,
  }));
}
